library/lua/gui.lua

```diff
@@ -476,12 +539,42 @@ end
 function View:onRenderBody(dc)
 end
 
+-- returns true if all of the following are true:
+-- - the view is not the focus owner
+-- - the view is not a descendent of the focus owner
+-- - the focus owner and all of its ancestors are visible and active
+local function should_send_input_to_focus_owner(view, focus_owner)
+    local iter = view
+    while iter do
+        if iter == focus_owner then
+            return false
```

Comment:
Perhaps the function name should be `send_input_to_focus_owner_from_view`, since it's determining if we should invoke the focus owner's `onInput()` function from the given view's `inputToSubviews()` function.

The first loop in this function prevents infinite loops when a focus owner has subviews of it's own. They are, by definition, part of the same focus group as the focus owner, so if the focus owner calls its own `inputToSubviews()` function, that input must be allowed to propagate normally instead of being sent *back* to the focus owner parent.

The second loop handles the case where the focus owner is not accessible because a parent is inactive or hidden. For example, say a widget on a Pages subview has focus, and the active Page is changed. Even though the (transitively) hidden widget still has the focus flag set to true, it should not be getting input until it is shown again.

Would it help if I added these examples to the code itself?